src/python/pants/core/util_rules/adhoc_binaries.py

```diff
@@ -0,0 +1,141 @@
+from pants.core.util_rules.system_binaries import SEARCH_PATHS, TarBinary
+from pants.engine.fs import DownloadFile
+from pants.engine.internals.native_engine import EMPTY_DIGEST, Digest, FileDigest
+from pants.engine.internals.selectors import Get
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class PythonBuildStandaloneBinary:
+    """A Python interpreter for use by `@rule` code as an alternative to BashBinary scripts.
+
+    This interpreter is provided by Python Build Standalone https://gregoryszorc.com/docs/python-build-standalone/main/,
+    which has a few caveats. Namely it doesn't play nicely with third-party sdists. Meaning Pants'
+    scripts being run by PythonBuildStandalone should avoid third-party sdists.
+    """
+
+    SYMLINK_DIRNAME = ".python-build-standalone"
+
+    _path: str
+    """The path to the Python binary inside the immutable input symlink."""
+    _digest: Digest
+
+    @property
+    def path(self) -> str:
+        return f"{PythonBuildStandaloneBinary.SYMLINK_DIRNAME}/{self._path}"
+
+    @property
+    def immutable_input_digests(self) -> FrozenDict[str, Digest]:
+        return FrozenDict({PythonBuildStandaloneBinary.SYMLINK_DIRNAME: self._digest})
+
+
+@rule(desc="Finding or downloading Python for scripts", level=LogLevel.TRACE)
+async def download_python_build_standalone(
+    platform: Platform, env_tgt: EnvironmentTarget, tar_binary: TarBinary
+) -> PythonBuildStandaloneBinary:
+    if isinstance(env_tgt.val, LocalEnvironmentTarget):
+        return PythonBuildStandaloneBinary(sys.executable, EMPTY_DIGEST)
+
+    url_plat, fingerprint, bytelen = {
+        # No PGO release for aarch64 it seems
+        "linux_arm64": (
+            "aarch64-unknown-linux-gnu-lto",
+            "c811df664ca233815f7b4bfec4648192b74599e2639554ba634d8891bb60a0ca",
+            154120974,
+        ),
+        "linux_x86_64": (
+            "x86_64-unknown-linux-gnu-pgo+lto",
+            "9cbad5f74dc958f65e7d0c3d159ba3ff5e1cab94b861d3f6e281c8a64fbbca48",
+            81272463,
+        ),
+        "macos_arm64": (
+            "aarch64-apple-darwin-pgo+lto",
+            "603af4d150bf3158961684e00104e273c0aab419b142760ff542aeee8bde32f9",
+            66451944,
+        ),
+        "macos_x86_64": (
+            "x86_64-apple-darwin-pgo+lto",
+            "5b4d4d425414cd4a8750b544ab05d284e5acd837d83f1c661320b8aea3762393",
+            65449500,
+        ),
+    }[platform.value]
+
+    filename = f"cpython-3.10.9+20230116-{url_plat}-full.tar.zst"
+    python_archive = await Get(
+        Digest,
+        DownloadFile(
+            f"https://github.com/thejcannon/python-build-standalone/releases/download/20230116/{filename}",
```

Comment:
I'd very much prefer this be forked and managed by the pantsbuild org :sweat_smile:

Reply:
Meh I suppose it's a new day and a new opportunity for a new intrinsic.

Reply:
But then again, no process cache for the intrinsic meaning we'd have to know the output's fingerprint.